cpp/src/cylon/indexing/slice.cpp

```diff
@@ -0,0 +1,170 @@
+                                         prefix_sum);
+  if (global_offset > total_length) {
+    return {Code::Invalid, "global offset exceeds total length of the dist table"};
+  }
+  // adjust global length if it exceeds total_length
+  if (global_offset + global_length > total_length) {
+    global_length = total_length - global_offset;
+  }
+
+  int64_t this_length = *(partition_lengths + rank);
+  assert(this_length == in->Rows());
+
+  int64_t local_offset = std::max(kZero, std::min(global_offset - prefix_sum, this_length));
+  int64_t local_length =
+      std::min(this_length, std::max(global_offset + global_length - prefix_sum, kZero))
+          - local_offset;
+
+  return Slice(in, local_offset, local_length, out);
+}
+
+Status DistributedSlice(const std::shared_ptr<Table> &in,
+                        int64_t offset,
+                        int64_t length,
+                        std::shared_ptr<Table> *out) {
+  return distributed_slice_impl(in, offset, length, nullptr, out);
+}
+
+/**
+ * Head the part of table to create a single table with specific number of rows
+ * @param tables, number of rows
+ * @return new table
+ */
+Status Head(const std::shared_ptr<Table> &table, int64_t num_rows,
+            std::shared_ptr<Table> *output) {
+  if (num_rows >= 0) {
+    return Slice(table, 0, num_rows, output);
+  } else
+    return {Code::Invalid, "Number of head rows should be >=0"};
+}
+
+Status DistributedHead(const std::shared_ptr<Table> &table, int64_t num_rows,
+                       std::shared_ptr<Table> *output) {
+
+  std::shared_ptr<arrow::Table> in_table = table->get_table();
+
+  if (num_rows >= 0) {
+    return distributed_slice_impl(table, 0, num_rows, nullptr, output);
+  } else {
+    return {Code::Invalid, "Number of head rows should be >=0"};
+  }
+}
+
+/**
+ * Tail the part of table to create a single table with specific number of rows
+ * @param tables, number of rows
+ * @return new table
+ */
+Status Tail(const std::shared_ptr<Table> &table, int64_t num_rows,
+            std::shared_ptr<Table> *output) {
+
+  std::shared_ptr<arrow::Table> in_table = table->get_table();
+  const int64_t table_size = in_table->num_rows();
+
+  if (num_rows >= 0) {
+    return Slice(table, table_size - num_rows, num_rows, output);
+  } else {
+    return {Code::Invalid, "Number of tailed rows should be >=0"};
+  }
+}
+
+Status DistributedTail(const std::shared_ptr<Table> &table, int64_t num_rows,
+                       std::shared_ptr<Table> *output) {
+  if (num_rows >= 0) {
+    const auto &ctx = table->GetContext();
+    std::shared_ptr<cylon::Column> partition_len_col;
+    const auto &num_row_scalar = Scalar::Make(arrow::MakeScalar(table->Rows()));
+    RETURN_CYLON_STATUS_IF_FAILED(ctx->GetCommunicator()
+                                      ->Allgather(num_row_scalar, &partition_len_col));
+    assert(ctx->GetWorldSize() == partition_len_col->length());
```

Comment:
Curious: So we limit the number of partitions to the number of processes? Why is that? Could it be multiples of number of processes?